File: Daily_Challenge/20230324_LC_2316.py
class Solution:
    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        # union find based method
        root = [i for i in range(n)]
        counts = [1] * n
        # path compression
        def find(x):
            if x != root[x]:
                root[x] = find(root[x])
            return root[x]
        # union
        def union(x, y):
            rx, ry = find(x), find(y)
            if rx != ry:
                root[ry] = rx
                counts[rx] += counts[ry]
        
        for v1, v2 in edges:
            union(v1, v2)
        ans = 0
        for i in range(n):
            ans += n - counts[find(i)]
        return ans // 2
        
        
        """
        # DFS based method
        # find the connected component
        # the answer is (sum(n - cc_size)) / 2
        graph = collections.defaultdict(list)
        for v1, v2 in edges:
            graph[v1].append(v2)
            graph[v2].append(v1)
        seen = [0] * n
        ans = 0
        # find the CCs size
        def dfs(node, seen):
            if seen[node]:
                return 0
            seen[node] = 1
            ans = 1
            for nxt in graph[node]:
                ans += dfs(nxt, seen)
            return ans
            
        
        for i in range(n):
            # this node is part of the previous calculated CCs
            if seen[i]:
                continue
            cur = dfs(i, seen) # find the CCs size
            ans += (n - cur) * cur
        return ans // 2
        
        """

File: Daily_Challenge/20230324_LC_2316.cpp
class Solution {
public:
    int find(int x, vector<int>& root){
        if (root[x] != x){
            root[x] = find(root[x], root);
        }
        return root[x];
    }

    void do_union(int x, int y, vector<int>& root, vector<int>& count){
        int rx = find(x, root), ry = find(y, root);
        if (rx != ry){
            root[ry] = rx;
            count[rx] += count[ry];
        }
    }
    void print_vector(vector<int>& arr){
        for (int i = 0; i < arr.size(); i++){
            cout << arr[i] << " ";
        }
        cout << endl;
    }
    long long countPairs(int n, vector<vector<int>>& edges) {
        vector<int> root(n, 0), count(n, 1);
        for (int i = 0; i < n; i++){
            root[i] = i;
        }
        for (vector<int> e: edges){
            do_union(e[0], e[1], root, count);
        }
        //print_vector(root);
        //print_vector(count);
        long long ans = 0;
       
        for (int i = 0; i < n; i++){
            int idx = find(i, root);
            ans += n - count[idx];
        }
        return ans/2;
    }
};